dbms/src/Storages/MergeTree/PKCondition.cpp

```diff
@@ -348,6 +353,47 @@ bool PKCondition::isPrimaryKeyPossiblyWrappedByMonotonicFunctionsImpl(
 }
 
 
+/// NOTE: Keep in the mind that such behavior could be incompatible inside ordinary expression.
+/// TODO: Use common methods for types conversions.
+static bool tryCastValueToType(const DataTypePtr & desired_type, const DataTypePtr & src_type, Field & src_value)
+{
+	if (desired_type->getName() == src_type->getName())
+		return true;
+
+	/// Try to correct type of constant for correct comparison
+	try
+	{
+		/// Convert String to Enum's value
+		if (auto data_type_enum = dynamic_cast<const IDataTypeEnum *>(desired_type.get()))
+		{
+			src_value = data_type_enum->castToValue(src_value);
+		}
+		/// Convert 'YYYY-MM-DD' Strings to Date
+		else if (typeid_cast<const DataTypeDate *>(desired_type.get()))
+		{
+			if (typeid_cast<const DataTypeString *>(src_type.get()))
+			{
+				src_value = stringToDateOrDateTime(src_value.safeGet<String>());
+			}
+		}
+		else if (desired_type->behavesAsNumber() && src_type->behavesAsNumber())
+		{
+			/// Ok, numeric types are mutually convertible
+		}
+		else
+		{
+			return false;
+		}
+	}
+	catch (...)
+	{
+		return false;
```

Comment:
There was old behaviour where exception is thrown immediately for comparison of Date/DateTime with string literal that contains invalid date. Is that behaviour has changed?